diff --git b/drivers/isdn/mISDN/l1oip_core.c a/drivers/isdn/mISDN/l1oip_core.c
index 072bb5e..b05022f 100644
--- b/drivers/isdn/mISDN/l1oip_core.c
+++ a/drivers/isdn/mISDN/l1oip_core.c
@@ -718,7 +718,8 @@ l1oip_socket_thread(void *data)
 		printk(KERN_DEBUG "%s: socket created and open\n",
 		       __func__);
 	while (!signal_pending(current)) {
-		iov_iter_kvec(&msg.msg_iter, READ, &iov, 1, recvbuf_size);
+		iov_iter_kvec(&msg.msg_iter, READ | ITER_KVEC, &iov, 1,
+				recvbuf_size);
 		recvlen = sock_recvmsg(socket, &msg, 0);
 		if (recvlen > 0) {
 			l1oip_socket_parse(hc, &sin_rx, recvbuf, recvlen);
