diff --git b/drivers/isdn/mISDN/socket.c a/drivers/isdn/mISDN/socket.c
index 09f2145..b6cc97e 100644
--- b/drivers/isdn/mISDN/socket.c
+++ a/drivers/isdn/mISDN/socket.c
@@ -563,7 +563,7 @@ done:
 
 static int
 data_sock_getname(struct socket *sock, struct sockaddr *addr,
-		  int peer)
+		  int *addr_len, int peer)
 {
 	struct sockaddr_mISDN	*maddr = (struct sockaddr_mISDN *) addr;
 	struct sock		*sk = sock->sk;
@@ -573,13 +573,14 @@ data_sock_getname(struct socket *sock, struct sockaddr *addr,
 
 	lock_sock(sk);
 
+	*addr_len = sizeof(*maddr);
 	maddr->family = AF_ISDN;
 	maddr->dev = _pms(sk)->dev->id;
 	maddr->channel = _pms(sk)->ch.nr;
 	maddr->sapi = _pms(sk)->ch.addr & 0xff;
 	maddr->tei = (_pms(sk)->ch.addr >> 8) & 0xff;
 	release_sock(sk);
-	return sizeof(*maddr);
+	return 0;
 }
 
 static const struct proto_ops data_sock_ops = {
