diff --git b/drivers/isdn/hardware/mISDN/hfcpci.c a/drivers/isdn/hardware/mISDN/hfcpci.c
index 2ead502..0de5735 100644
--- b/drivers/isdn/hardware/mISDN/hfcpci.c
+++ a/drivers/isdn/hardware/mISDN/hfcpci.c
@@ -2270,7 +2270,7 @@ static struct pci_driver hfc_driver = {
 };
 
 static int
-_hfcpci_softirq(struct device *dev, void *unused)
+_hfcpci_softirq(struct device *dev, void *arg)
 {
 	struct hfc_pci  *hc = dev_get_drvdata(dev);
 	struct bchannel *bch;
@@ -2295,9 +2295,9 @@ _hfcpci_softirq(struct device *dev, void *unused)
 }
 
 static void
-hfcpci_softirq(struct timer_list *unused)
+hfcpci_softirq(void *arg)
 {
-	WARN_ON_ONCE(driver_for_each_device(&hfc_driver.driver, NULL, NULL,
+	WARN_ON_ONCE(driver_for_each_device(&hfc_driver.driver, NULL, arg,
 				      _hfcpci_softirq) != 0);
 
 	/* if next event would be in the past ... */
@@ -2332,7 +2332,9 @@ HFC_init(void)
 	if (poll != HFCPCI_BTRANS_THRESHOLD) {
 		printk(KERN_INFO "%s: Using alternative poll value of %d\n",
 		       __func__, poll);
-		timer_setup(&hfc_tl, hfcpci_softirq, 0);
+		hfc_tl.function = (void *)hfcpci_softirq;
+		hfc_tl.data = 0;
+		init_timer(&hfc_tl);
 		hfc_tl.expires = jiffies + tics;
 		hfc_jiffies = hfc_tl.expires;
 		add_timer(&hfc_tl);
