diff --git b/drivers/isdn/mISDN/fsm.c a/drivers/isdn/mISDN/fsm.c
index 9a8d08d..cabcb90 100644
--- b/drivers/isdn/mISDN/fsm.c
+++ a/drivers/isdn/mISDN/fsm.c
@@ -32,10 +32,8 @@ mISDN_FsmNew(struct Fsm *fsm,
 {
 	int i;
 
-	fsm->jumpmatrix =
-		kzalloc(array3_size(sizeof(FSMFNPTR), fsm->state_count,
-				    fsm->event_count),
-			GFP_KERNEL);
+	fsm->jumpmatrix = kzalloc(sizeof(FSMFNPTR) * fsm->state_count *
+				  fsm->event_count, GFP_KERNEL);
 	if (fsm->jumpmatrix == NULL)
 		return -ENOMEM;
 
