diff --git b/drivers/isdn/hardware/mISDN/hfcpci.c a/drivers/isdn/hardware/mISDN/hfcpci.c
index 763de33..f4170a9 100644
--- b/drivers/isdn/hardware/mISDN/hfcpci.c
+++ a/drivers/isdn/hardware/mISDN/hfcpci.c
@@ -301,9 +301,8 @@ reset_hfcpci(struct hfc_pci *hc)
  * Timer function called when kernel timer expires
  */
 static void
-hfcpci_Timer(struct timer_list *t)
+hfcpci_Timer(struct hfc_pci *hc)
 {
-	struct hfc_pci *hc = from_timer(hc, t, hw.timer);
 	hc->hw.timer.expires = jiffies + 75;
 	/* WD RESET */
 /*
@@ -2045,7 +2044,7 @@ setup_hw(struct hfc_pci *hc)
 	Write_hfc(hc, HFCPCI_INT_M1, hc->hw.int_m1);
 	/* At this point the needed PCI config is done */
 	/* fifos are still not enabled */
-	timer_setup(&hc->hw.timer, hfcpci_Timer, 0);
+	setup_timer(&hc->hw.timer, (void *)hfcpci_Timer, (long)hc);
 	/* default PCM master */
 	test_and_set_bit(HFC_CFG_MASTER, &hc->cfg);
 	return 0;
